src/persist-client/src/usage.rs

```diff
@@ -122,19 +386,121 @@ impl StorageUsageClient {
             .await?;
         Ok(total_size)
     }
+}
 
-    #[cfg(test)]
-    fn open_from_blob(blob: Arc<dyn Blob + Send + Sync>) -> StorageUsageClient {
-        use mz_build_info::DUMMY_BUILD_INFO;
-        use mz_ore::metrics::MetricsRegistry;
-        use mz_ore::now::SYSTEM_TIME;
+#[derive(Debug)]
+struct ShardUsageCumulativeMaybeRacy {
+    current_state_batches_bytes: u64,
+    current_state_bytes: u64,
+    referenced_bytes: u64,
+    not_leaked_bytes: u64,
+    total_bytes: u64,
+}
 
-        use crate::PersistConfig;
-        let cfg = PersistConfig::new(&DUMMY_BUILD_INFO, SYSTEM_TIME.clone());
-        StorageUsageClient {
-            blob: Arc::clone(&blob),
-            metrics: Arc::new(Metrics::new(&cfg, &MetricsRegistry::new())),
+impl From<ShardUsageCumulativeMaybeRacy> for ShardUsage {
+    fn from(x: ShardUsageCumulativeMaybeRacy) -> Self {
+        let leaked_bytes = x
+            .total_bytes
+            .checked_sub(x.not_leaked_bytes)
+            .expect("blob inputs should be cumulative");
+        let mut referenced_not_current_state_bytes = x
+            .referenced_bytes
+            .checked_sub(x.current_state_bytes)
+            .expect("state inputs should be cumulative");
+        let mut current_state_rollups_bytes = x
+            .current_state_bytes
+            .checked_sub(x.current_state_batches_bytes)
+            .expect("state inputs should be cumulative");
+        let mut current_state_batches_bytes = x.current_state_batches_bytes;
+
+        // If we could transactionally read both blob and consensus, the
+        // cumulative numbers would all line up. We can't, so we have to adjust
+        // them up a bit to account for the race condition. If this happens,
+        // numbers will only get smaller (by policy, we prefer to under-count
+        // for billing).
```

Comment:
would love a bit more detail on the "what" of the adjustment here, this bit is also subtle